src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs

```diff
@@ -710,8 +711,12 @@ internal void Add(string? ns, object o, TempAssembly assembly)
                 TempAssembly? tempAssembly;
                 if (_cache.TryGetValue(key, out tempAssembly) && tempAssembly == assembly)
                     return;
-                Dictionary<TempAssemblyCacheKey, TempAssembly> _copy = new Dictionary<TempAssemblyCacheKey, TempAssembly>(_cache); // clone
-                _copy[key] = assembly;
+                ConditionalWeakTable<TempAssemblyCacheKey, TempAssembly> _copy = new ConditionalWeakTable<TempAssemblyCacheKey, TempAssembly>(); // clone
+                foreach (KeyValuePair<TempAssemblyCacheKey, TempAssembly> kvp in _cache)
+                {
+                    _copy.Add(kvp.Key, kvp.Value);
+                }
+                _copy.Add(key, assembly);
```

Comment:
The `key` instance being constructed here isn't rooted by anything else.  Doesn't this change mean that things added to the cache will end up being collected the next time the GC runs?  Is that what's desired?

Reply:
@stephentoub you are correct. The ConditionalWeakTable needs to use the Type object as the key, and the value probably needs to be a `Dictionary<TempAssemblyCacheKey, TempAssembly>()`. So we first lookup by Type, and then in the returned dictionary, do the lookup by TempAssemblyCacheKey. We could use a list, but then we introduce the edge case of if someone has code to create a serializer with arbitrary namespaces it could slow down the lookup.  
If this goes to servicing instead of initial release, I think it might be worth having two caches, one for types where the ALC is not collectible, and one for types where the ALC is collectible. ALC has an IsCollectible property. I don't know what the overhead is for ConditionalWeakTable, but I suspect it's quite a lot higher than it is for Dictionary.